### PR TITLE
Make BArray use an actual destructor

### PR DESCRIPTION
**Pull request by thewilsonator**
The makes the interface more similar between backend.barray and
root.array

### CODE REVIEW

**Comment by WalterBright** [comment]
Not using destructors until everything is `nothrow` because of code bloat generated for it.

**Comment by thewilsonator** [comment]
Ahh, good point. are exceptions even used in the backend?

**Comment by WalterBright** [comment]
No, they're not used. But the compiler doesn't know that, because it's not `nothrow`.

**Comment by thewilsonator** [comment]
Lets change that then!

**Comment by thewilsonator** [comment]
https://github.com/dlang/dmd/pull/9975

**Comment by thewilsonator** [comment]
Reopening so I don't forget about it.

File: src/dmd/backend/barray.d [hunk 1]
```diff
@@ -110,6 +110,4 @@ unittest
         v = cast(int) i * 2;
     foreach (i, ref const v; a[])
         assert(v == i * 2);
-    a.dtor();
-    assert(a.length == 0);
```
Comment (jacob-carlborg) [inline src/dmd/backend/barray.d]: Why was this assertion removed?

Comment (thewilsonator) [reply src/dmd/backend/barray.d]: because you can't assert on effects of the destructors, only side effects such as writing to globals.

**Review by jacob-carlborg — APPROVED** [review]

**Pull request merged**